File: java/src/main/java/org/libpetri/runtime/PrecompiledNet.java
package org.libpetri.runtime;

import java.util.*;

import org.libpetri.core.*;

/**
 * Precompiled flat-array net representation of a {@link PetriNet} for the {@link PrecompiledNetExecutor}.
 *
 * <p>Compiles the net topology into flat arrays and operation sequences that eliminate
 * virtual dispatch and sealed-type pattern matching from the hot path.
 *
 * <h2>Consume Operations</h2>
 * <p>Each transition's input/reset arcs are compiled into a flat {@code int[]} of opcodes:
 * <pre>
 *   CONSUME_ONE(0)     placeId
 *   CONSUME_N(1)       placeId count
 *   CONSUME_ALL(2)     placeId
 *   CONSUME_ATLEAST(3) placeId minimum
 *   RESET(4)           placeId
 * </pre>
 *
 * <h2>Priority Levels</h2>
 * <p>Distinct priority values are pre-sorted descending and indexed. Each transition
 * maps to a priority index for O(1) ready-queue insertion.
 *
 * @see PrecompiledNetExecutor
 * @see CompiledNet
 */
public final class PrecompiledNet {

    // Consume operation opcodes
    static final int CONSUME_ONE     = 0;
    static final int CONSUME_N       = 1;
    static final int CONSUME_ALL     = 2;
    static final int CONSUME_ATLEAST = 3;
    static final int RESET           = 4;

    // Net identity
    final String netName;

    // Net topology
    final int placeCount;
    final int transitionCount;
    final int wordCount;

    // ID mappings
    final Place<?>[] placesById;
    final Transition[] transitionsById;
    final Map<Place<?>, Integer> placeIndex;
    final IdentityHashMap<Transition, Integer> transitionIndex;

    // Operation sequences per transition
    final int[][] consumeOps;   // consume/reset opcodes
    final int[][] readOps;      // read-arc place IDs

    // Enablement masks (reused from CompiledNet)
    final long[][] needsMaskWords;
    final long[][] inhibitorMaskWords;
    final CompiledNet.CardinalityCheck[] cardinalityChecks;

    // Reverse index
    final int[][] placeToTransitions;

    // Consumption place IDs per transition (input + reset places, deduplicated)
    final int[][] consumptionPlaceIds;

    // Timing (precomputed)
    final long[] earliestNanos;
    final long[] latestNanos;
    final long[] earliestMillis;  // precomputed to avoid Division on hot path
    final long[] latestMillis;    // precomputed to avoid virtual dispatch on hot path
    final boolean[] hasDeadline;

    // Priority
    final int[] priorities;
    final int[] transitionToPriorityIndex; // tid → index into priorityLevels
    final int[] priorityLevels;            // distinct values, sorted descending
    final int distinctPriorityCount;

    // Global flags
    final boolean allImmediate;
    final boolean allSamePriority;
    final boolean anyDeadlines;

    // Output validation: precomputed for fast path
    // For Out.Place specs, stores the single expected place ID; -1 for complex specs; -2 for no spec
    final int[] simpleOutputPlaceId;

    // Input place count per transition (for pre-sizing TokenInput)
    final int[] inputPlaceCount;

    // Input place masks for reset-clock detection (bitmask per transition)
    final long[][] inputPlaceMaskWords;

    // Sparse enablement: precomputed for O(1) single-input checks
    // -2 = empty (no inputs), -1 = multi-word, >= 0 = single word index
    final int[] needsSingleWordIndex;
    final long[] needsSingleWordMask;
    final int[][] needsSparseIndices;   // non-null only for multi-word
    final long[][] needsSparseMasks;

    final int[] inhibitorSingleWordIndex;  // -2 = no inhibitors (skip entirely)
    final long[] inhibitorSingleWordMask;
    final int[][] inhibitorSparseIndices;
    final long[][] inhibitorSparseMasks;

    private PrecompiledNet(CompiledNet compiled) {
        this.netName = compiled.net().name();
        this.placeCount = compiled.placeCount();
        this.transitionCount = compiled.transitionCount();
        this.wordCount = compiled.wordCount();

        // Copy ID mappings from CompiledNet
        this.placesById = new Place<?>[placeCount];
        for (int i = 0; i < placeCount; i++) placesById[i] = compiled.place(i);

        this.transitionsById = new Transition[transitionCount];
        this.transitionIndex = new IdentityHashMap<>(transitionCount * 2);
        for (int i = 0; i < transitionCount; i++) {
            transitionsById[i] = compiled.transition(i);
            transitionIndex.put(transitionsById[i], i);
        }

        this.placeIndex = new HashMap<>(placeCount * 2);
        for (int i = 0; i < placeCount; i++) {
            placeIndex.put(placesById[i], i);
        }

        // Reuse enablement masks
        this.needsMaskWords = new long[transitionCount][];
        this.inhibitorMaskWords = new long[transitionCount][];
        this.cardinalityChecks = new CompiledNet.CardinalityCheck[transitionCount];
        for (int tid = 0; tid < transitionCount; tid++) {
            needsMaskWords[tid] = compiled.needsMask(tid).toLongArray();
            inhibitorMaskWords[tid] = compiled.inhibitorMask(tid).toLongArray();
            cardinalityChecks[tid] = compiled.cardinalityCheck(tid);
        }

        // Reuse reverse index and consumption IDs
        this.placeToTransitions = new int[placeCount][];
        for (int pid = 0; pid < placeCount; pid++) {
            placeToTransitions[pid] = compiled.affectedTransitions(pid);
        }

        this.consumptionPlaceIds = new int[transitionCount][];
        for (int tid = 0; tid < transitionCount; tid++) {
            consumptionPlaceIds[tid] = compiled.consumptionPlaceIds(tid);
        }

        // Compile sparse enablement masks
        this.needsSingleWordIndex = new int[transitionCount];
        this.needsSingleWordMask = new long[transitionCount];
        this.needsSparseIndices = new int[transitionCount][];
        this.needsSparseMasks = new long[transitionCount][];
        this.inhibitorSingleWordIndex = new int[transitionCount];
        this.inhibitorSingleWordMask = new long[transitionCount];
        this.inhibitorSparseIndices = new int[transitionCount][];
        this.inhibitorSparseMasks = new long[transitionCount][];

        for (int tid = 0; tid < transitionCount; tid++) {
            compileSparse(needsMaskWords[tid], needsSingleWordIndex, needsSingleWordMask,
                          needsSparseIndices, needsSparseMasks, tid);
            compileSparse(inhibitorMaskWords[tid], inhibitorSingleWordIndex, inhibitorSingleWordMask,
                          inhibitorSparseIndices, inhibitorSparseMasks, tid);
        }

        // Compile consume/read operation sequences
        this.consumeOps = new int[transitionCount][];
        this.readOps = new int[transitionCount][];
        this.inputPlaceMaskWords = new long[transitionCount][];

        this.simpleOutputPlaceId = new int[transitionCount];
        this.inputPlaceCount = new int[transitionCount];

        for (int tid = 0; tid < transitionCount; tid++) {
            Transition t = transitionsById[tid];
            consumeOps[tid] = compileFireProgram(t);
            readOps[tid] = compileReadProgram(t);
            inputPlaceMaskWords[tid] = compileInputMask(t);

            // Precompute input place count (for TokenInput pre-sizing)
            inputPlaceCount[tid] = t.inputSpecs().size() + t.reads().size();

            // Precompute output validation fast path
            if (t.outputSpec() == null) {
                simpleOutputPlaceId[tid] = -2; // no spec
            } else if (t.outputSpec() instanceof Arc.Out.Place p) {
                simpleOutputPlaceId[tid] = placeIndex.get(p.place());
            } else {
                simpleOutputPlaceId[tid] = -1; // complex spec, use full validation
            }
        }

        // Precompute timing
        this.earliestNanos = new long[transitionCount];
        this.latestNanos = new long[transitionCount];
        this.earliestMillis = new long[transitionCount];
        this.latestMillis = new long[transitionCount];
        this.hasDeadline = new boolean[transitionCount];
        boolean anyDl = false;
        boolean allImm = true;

        for (int tid = 0; tid < transitionCount; tid++) {
            Transition t = transitionsById[tid];
            earliestNanos[tid] = t.timing().earliest().toNanos();
            earliestMillis[tid] = t.timing().earliest().toMillis();
            latestNanos[tid] = t.timing().hasDeadline()
                ? t.timing().latest().toNanos()
                : Long.MAX_VALUE;
            latestMillis[tid] = t.timing().hasDeadline()
                ? t.timing().latest().toMillis()
                : Long.MAX_VALUE;
            hasDeadline[tid] = t.timing().hasDeadline();
            if (hasDeadline[tid]) anyDl = true;
            if (!(t.timing() instanceof Timing.Immediate) &&
                !(t.timing() instanceof Timing.Unconstrained)) {
                allImm = false;
            }
        }
        this.anyDeadlines = anyDl;
        this.allImmediate = allImm;

        // Precompute priorities
        this.priorities = new int[transitionCount];
        var distinctPrios = new TreeSet<Integer>(Comparator.reverseOrder());
        for (int tid = 0; tid < transitionCount; tid++) {
            priorities[tid] = transitionsById[tid].priority();
            distinctPrios.add(priorities[tid]);
        }
        this.priorityLevels = distinctPrios.stream().mapToInt(Integer::intValue).toArray();
        this.distinctPriorityCount = priorityLevels.length;
        this.allSamePriority = distinctPriorityCount <= 1;

        // Map each transition to its priority index
        var prioToIndex = new HashMap<Integer, Integer>();
        for (int i = 0; i < priorityLevels.length; i++) {
            prioToIndex.put(priorityLevels[i], i);
        }
        this.transitionToPriorityIndex = new int[transitionCount];
        for (int tid = 0; tid < transitionCount; tid++) {
            transitionToPriorityIndex[tid] = prioToIndex.get(priorities[tid]);
        }
    }

    /**
     * Compiles a PetriNet into a PrecompiledNet.
     *
     * @param net the symbolic Petri net
     * @return precompiled net ready for execution
     */
    public static PrecompiledNet compile(PetriNet net) {
        return new PrecompiledNet(CompiledNet.compile(net));
    }

    /**
     * Compiles using an existing CompiledNet to reuse its masks and indices.
     *
     * @param compiled pre-compiled net
     * @return precompiled net ready for execution
     */
    public static PrecompiledNet compile(CompiledNet compiled) {
        return new PrecompiledNet(compiled);
    }

    // ==================== Operation Sequence Compilation ====================

    private int[] compileFireProgram(Transition t) {
        // Estimate size: each input spec needs 2-3 ints, each reset needs 2
        var ops = new ArrayList<Integer>();

        for (var in : t.inputSpecs()) {
            int pid = placeIndex.get(in.place());
            switch (in) {
                case Arc.In.One _ -> {
                    ops.add(CONSUME_ONE);
                    ops.add(pid);
                }
                case Arc.In.Exactly e -> {
                    ops.add(CONSUME_N);
                    ops.add(pid);
                    ops.add(e.count());
                }
                case Arc.In.All _ -> {
                    ops.add(CONSUME_ALL);
                    ops.add(pid);
                }
                case Arc.In.AtLeast a -> {
                    ops.add(CONSUME_ATLEAST);
                    ops.add(pid);
                    ops.add(a.minimum());
                }
            }
        }

        for (var arc : t.resets()) {
            int pid = placeIndex.get(arc.place());
            ops.add(RESET);
            ops.add(pid);
        }

        int[] result = new int[ops.size()];
        for (int i = 0; i < result.length; i++) result[i] = ops.get(i);
        return result;
    }

    private int[] compileReadProgram(Transition t) {
        var reads = t.reads();
        int[] result = new int[reads.size()];
        for (int i = 0; i < result.length; i++) {
            result[i] = placeIndex.get(reads.get(i).place());
        }
        return result;
    }

    private long[] compileInputMask(Transition t) {
        var bits = new BitSet(placeCount);
        for (var in : t.inputSpecs()) {
            bits.set(placeIndex.get(in.place()));
        }
        return bits.toLongArray();
    }

    // ==================== Sparse Mask Compilation ====================

    private static void compileSparse(long[] maskWords,
                                      int[] singleWordIndex, long[] singleWordMask,
                                      int[][] sparseIndices, long[][] sparseMasks, int tid) {
        // Count non-zero words
        int nonZeroCount = 0;
        int lastNonZeroWord = -1;
        for (int w = 0; w < maskWords.length; w++) {
            if (maskWords[w] != 0) {
                nonZeroCount++;
                lastNonZeroWord = w;
            }
        }

        if (nonZeroCount == 0) {
            singleWordIndex[tid] = -2; // empty
        } else if (nonZeroCount == 1) {
            singleWordIndex[tid] = lastNonZeroWord;
            singleWordMask[tid] = maskWords[lastNonZeroWord];
        } else {
            singleWordIndex[tid] = -1; // multi-word
            int[] indices = new int[nonZeroCount];
            long[] masks = new long[nonZeroCount];
            int idx = 0;
            for (int w = 0; w < maskWords.length; w++) {
                if (maskWords[w] != 0) {
                    indices[idx] = w;
                    masks[idx] = maskWords[w];
                    idx++;
                }
            }
            sparseIndices[tid] = indices;
            sparseMasks[tid] = masks;
        }
    }

    // ==================== Enablement Check ====================

    boolean canEnableBitmap(int tid, long[] markingSnapshot) {
        // Needs check (containsAll)
        int nwi = needsSingleWordIndex[tid];
        if (nwi == -1) {
            // Multi-word sparse
            int[] indices = needsSparseIndices[tid];
            long[] masks = needsSparseMasks[tid];
            for (int i = 0; i < indices.length; i++) {
                int w = indices[i];
                if (w >= markingSnapshot.length || (markingSnapshot[w] & masks[i]) != masks[i])
                    return false;
            }
        } else if (nwi >= 0) {
            // Single-word
            if (nwi >= markingSnapshot.length ||
                (markingSnapshot[nwi] & needsSingleWordMask[tid]) != needsSingleWordMask[tid])
                return false;
        }
        // nwi == -2: empty needs mask, always passes

        // Inhibitor check (intersects)
        int iwi = inhibitorSingleWordIndex[tid];
        if (iwi == -2) return true; // no inhibitors
        if (iwi >= 0) {
            // Single-word
            return iwi >= markingSnapshot.length ||
                   (markingSnapshot[iwi] & inhibitorSingleWordMask[tid]) == 0;
        }
        // Multi-word sparse
        int[] indices = inhibitorSparseIndices[tid];
        long[] masks = inhibitorSparseMasks[tid];
        for (int i = 0; i < indices.length; i++) {
            int w = indices[i];
            if (w < markingSnapshot.length && (markingSnapshot[w] & masks[i]) != 0)
                return false;
        }
        return true;
    }

    // ==================== Accessors ====================

    public Place<?> place(int pid) { return placesById[pid]; }
    public Transition transition(int tid) { return transitionsById[tid]; }

    public int placeId(Place<?> place) {
        Integer id = placeIndex.get(place);
        if (id == null) throw new IllegalArgumentException("Unknown place: " + place.name());
        return id;
    }

    public int transitionId(Transition t) {
        Integer id = transitionIndex.get(t);
        if (id == null) throw new IllegalArgumentException("Unknown transition: " + t.name());
        return id;
    }
}

File: typescript/src/runtime/precompiled-net.ts
/**
 * @module precompiled-net
 *
 * Flat-array precompiled representation of a PetriNet for high-performance execution.
 *
 * **Design**: Compiles from a `CompiledNet`, converting all per-transition data into
 * flat typed arrays indexed by transition ID (tid) or place ID (pid). Eliminates Map
 * lookups and object traversals from the hot path.
 *
 * **Sparse enablement**: Each transition's needs/inhibitor masks are classified as
 * empty (-2), single-word (>=0), or multi-word (-1). Single-word masks avoid inner
 * loops entirely; multi-word uses precomputed sparse indices to skip zero words.
 *
 * **Opcode programs**: Input consumption is compiled to compact opcode sequences
 * (CONSUME_ONE, CONSUME_N, CONSUME_ALL, CONSUME_ATLEAST, RESET) that the executor
 * dispatches without inspecting In spec objects.
 *
 * @see CompiledNet for the base bitmap representation
 * @see PrecompiledNetExecutor for the executor that uses this representation
 */
import type { PetriNet } from '../core/petri-net.js';
import type { Place } from '../core/place.js';
import type { Transition } from '../core/transition.js';
import type { Out } from '../core/out.js';
import { CompiledNet, WORD_SHIFT, BIT_MASK } from './compiled-net.js';
import type { CardinalityCheck } from './compiled-net.js';
import { earliest as timingEarliest, latest as timingLatest, hasDeadline as timingHasDeadline } from '../core/timing.js';

// ==================== Opcodes ====================

/** Consume exactly 1 token from place. Next word: pid. */
export const CONSUME_ONE = 0;
/** Consume exactly N tokens. Next words: pid, count. */
export const CONSUME_N = 1;
/** Consume all tokens. Next word: pid. */
export const CONSUME_ALL = 2;
/** Consume at-least N tokens (consumes all). Next words: pid, minimum. */
export const CONSUME_ATLEAST = 3;
/** Reset (clear all tokens). Next word: pid. */
export const RESET = 4;

/** Sparse mask classification: no bits set. */
const SPARSE_EMPTY = -2;
/** Sparse mask classification: bits span multiple words. */
const SPARSE_MULTI = -1;

/**
 * Immutable precompiled representation of a Petri net.
 *
 * All data is stored in flat typed arrays indexed by tid/pid for cache-friendly,
 * branch-free access on the hot path. Compiled once, reused across executions.
 */
export class PrecompiledNet {
  // ==================== Identity ====================
  readonly compiled: CompiledNet;
  readonly placeCount: number;
  readonly transitionCount: number;
  readonly wordCount: number;

  // ==================== Place Cache ====================
  /** Places indexed by pid — avoids compiled.place(pid) indirection on hot path. */
  readonly places: readonly Place<any>[];

  // ==================== Opcode Programs ====================
  /** Per-transition consume opcode sequences. */
  readonly consumeOps: readonly (readonly number[])[];
  /** Per-transition read-arc place IDs. */
  readonly readOps: readonly (readonly number[])[];

  // ==================== Enablement Masks ====================
  readonly needsMask: readonly Uint32Array[];
  readonly inhibitorMask: readonly Uint32Array[];

  // ==================== Sparse Enablement (PERF-042) ====================
  readonly needsSingleWordIndex: Int8Array;
  readonly needsSingleWordMask: Uint32Array;
  readonly needsSparseIndices: readonly (readonly number[])[];
  readonly needsSparseMasks: readonly (readonly number[])[];

  readonly inhibitorSingleWordIndex: Int8Array;
  readonly inhibitorSingleWordMask: Uint32Array;
  readonly inhibitorSparseIndices: readonly (readonly number[])[];
  readonly inhibitorSparseMasks: readonly (readonly number[])[];

  // ==================== Timing (CONC-024) ====================
  readonly earliestMs: Float64Array;
  readonly latestMs: Float64Array;
  readonly hasDeadline: Uint8Array;

  // ==================== Priority (CONC-023) ====================
  readonly priorities: Int32Array;
  readonly transitionToPriorityIndex: Uint32Array;
  readonly priorityLevels: readonly number[];
  readonly distinctPriorityCount: number;

  // ==================== Output Fast Path ====================
  /** -2=no spec, -1=complex, >=0=single output place ID. */
  readonly simpleOutputPlaceId: Int32Array;

  // ==================== Input Precomputation ====================
  readonly inputPlaceCount: Uint32Array;
  readonly inputPlaceMaskWords: readonly Uint32Array[];

  // ==================== Reverse Index ====================
  readonly placeToTransitions: readonly (readonly number[])[];
  readonly consumptionPlaceIds: readonly (readonly number[])[];

  // ==================== Cardinality & Guards ====================
  readonly cardinalityChecks: readonly (CardinalityCheck | null)[];
  readonly hasGuards: readonly boolean[];

  // ==================== Global Flags ====================
  readonly allImmediate: boolean;
  readonly allSamePriority: boolean;
  readonly anyDeadlines: boolean;

  private constructor(compiled: CompiledNet) {
    this.compiled = compiled;
    this.placeCount = compiled.placeCount;
    this.transitionCount = compiled.transitionCount;
    this.wordCount = compiled.wordCount;

    const tc = compiled.transitionCount;
    const wc = compiled.wordCount;

    // Cache place references
    const places: Place<any>[] = new Array(this.placeCount);
    for (let pid = 0; pid < this.placeCount; pid++) {
      places[pid] = compiled.place(pid);
    }
    this.places = places;

    // Copy masks from CompiledNet
    const needsMask: Uint32Array[] = new Array(tc);
    const inhibitorMask: Uint32Array[] = new Array(tc);
    for (let tid = 0; tid < tc; tid++) {
      // Access via canEnableBitmap internals — copy the mask arrays
      // We need to extract them; use the CompiledNet's accessor pattern
      needsMask[tid] = new Uint32Array(wc);
      inhibitorMask[tid] = new Uint32Array(wc);
    }

    // Build needs/inhibitor masks by probing CompiledNet
    // We reconstruct from transition arc specs since CompiledNet doesn't expose raw masks
    for (let tid = 0; tid < tc; tid++) {
      const t = compiled.transition(tid);
      const needs = needsMask[tid]!;
      const inhibitors = inhibitorMask[tid]!;

      for (const inSpec of t.inputSpecs) {
        const pid = compiled.placeId(inSpec.place);
        needs[pid >>> WORD_SHIFT]! |= (1 << (pid & BIT_MASK));
      }
      for (const arc of t.reads) {
        const pid = compiled.placeId(arc.place);
        needs[pid >>> WORD_SHIFT]! |= (1 << (pid & BIT_MASK));
      }
      for (const arc of t.inhibitors) {
        const pid = compiled.placeId(arc.place);
        inhibitors[pid >>> WORD_SHIFT]! |= (1 << (pid & BIT_MASK));
      }
    }
    this.needsMask = needsMask;
    this.inhibitorMask = inhibitorMask;

    // ==================== Sparse Enablement ====================
    this.needsSingleWordIndex = new Int8Array(tc);
    this.needsSingleWordMask = new Uint32Array(tc);
    const needsSparseIndices: number[][] = new Array(tc);
    const needsSparseMasks: number[][] = new Array(tc);

    this.inhibitorSingleWordIndex = new Int8Array(tc);
    this.inhibitorSingleWordMask = new Uint32Array(tc);
    const inhibitorSparseIndices: number[][] = new Array(tc);
    const inhibitorSparseMasks: number[][] = new Array(tc);

    for (let tid = 0; tid < tc; tid++) {
      compileSparse(needsMask[tid]!, wc,
        this.needsSingleWordIndex, this.needsSingleWordMask,
        needsSparseIndices, needsSparseMasks, tid);
      compileSparse(inhibitorMask[tid]!, wc,
        this.inhibitorSingleWordIndex, this.inhibitorSingleWordMask,
        inhibitorSparseIndices, inhibitorSparseMasks, tid);
    }
    this.needsSparseIndices = needsSparseIndices;
    this.needsSparseMasks = needsSparseMasks;
    this.inhibitorSparseIndices = inhibitorSparseIndices;
    this.inhibitorSparseMasks = inhibitorSparseMasks;

    // ==================== Opcode Programs ====================
    const consumeOps: number[][] = new Array(tc);
    const readOps: number[][] = new Array(tc);
    for (let tid = 0; tid < tc; tid++) {
      const t = compiled.transition(tid);
      consumeOps[tid] = compileConsumeProgram(t, compiled);
      readOps[tid] = compileReadProgram(t, compiled);
    }
    this.consumeOps = consumeOps;
    this.readOps = readOps;

    // ==================== Reverse Index & Consumption ====================
    const placeToTransitions: number[][] = new Array(this.placeCount);
    const consumptionPlaceIds: number[][] = new Array(tc);
    for (let pid = 0; pid < this.placeCount; pid++) {
      placeToTransitions[pid] = [...compiled.affectedTransitions(pid)];
    }
    for (let tid = 0; tid < tc; tid++) {
      consumptionPlaceIds[tid] = [...compiled.consumptionPlaceIds(tid)];
    }
    this.placeToTransitions = placeToTransitions;
    this.consumptionPlaceIds = consumptionPlaceIds;

    // ==================== Cardinality & Guards ====================
    const cardinalityChecks: (CardinalityCheck | null)[] = new Array(tc);
    const hasGuards: boolean[] = new Array(tc);
    for (let tid = 0; tid < tc; tid++) {
      cardinalityChecks[tid] = compiled.cardinalityCheck(tid);
      hasGuards[tid] = compiled.hasGuards(tid);
    }
    this.cardinalityChecks = cardinalityChecks;
    this.hasGuards = hasGuards;

    // ==================== Timing ====================
    this.earliestMs = new Float64Array(tc);
    this.latestMs = new Float64Array(tc);
    this.hasDeadline = new Uint8Array(tc);
    let anyDeadlines = false;
    let allImm = true;

    for (let tid = 0; tid < tc; tid++) {
      const t = compiled.transition(tid);
      this.earliestMs[tid] = timingEarliest(t.timing);
      this.latestMs[tid] = timingLatest(t.timing);
      if (timingHasDeadline(t.timing)) {
        this.hasDeadline[tid] = 1;
        anyDeadlines = true;
      }
      if (t.timing.type !== 'immediate') allImm = false;
    }
    this.anyDeadlines = anyDeadlines;
    this.allImmediate = allImm;

    // ==================== Priority ====================
    this.priorities = new Int32Array(tc);
    const prioritySet = new Set<number>();
    const firstPriority = tc > 0 ? compiled.transition(0).priority : 0;
    let samePrio = true;

    for (let tid = 0; tid < tc; tid++) {
      const p = compiled.transition(tid).priority;
      this.priorities[tid] = p;
      prioritySet.add(p);
      if (p !== firstPriority) samePrio = false;
    }
    this.allSamePriority = samePrio;

    // Sort priorities descending
    const levels = [...prioritySet].sort((a, b) => b - a);
    this.priorityLevels = levels;
    this.distinctPriorityCount = levels.length;

    // Map tid -> priority queue index
    this.transitionToPriorityIndex = new Uint32Array(tc);
    const levelIndex = new Map<number, number>();
    for (let i = 0; i < levels.length; i++) {
      levelIndex.set(levels[i]!, i);
    }
    for (let tid = 0; tid < tc; tid++) {
      this.transitionToPriorityIndex[tid] = levelIndex.get(this.priorities[tid]!)!;
    }

    // ==================== Output Fast Path ====================
    this.simpleOutputPlaceId = new Int32Array(tc);
    for (let tid = 0; tid < tc; tid++) {
      const t = compiled.transition(tid);
      if (t.outputSpec === null) {
        this.simpleOutputPlaceId[tid] = -2; // no spec
      } else {
        const simplePid = simpleOutputPlace(t.outputSpec, compiled);
        this.simpleOutputPlaceId[tid] = simplePid;
      }
    }

    // ==================== Input Precomputation ====================
    this.inputPlaceCount = new Uint32Array(tc);
    const inputPlaceMaskWords: Uint32Array[] = new Array(tc);
    for (let tid = 0; tid < tc; tid++) {
      const t = compiled.transition(tid);
      this.inputPlaceCount[tid] = t.inputSpecs.length + t.reads.length;
      const mask = new Uint32Array(wc);
      for (const spec of t.inputSpecs) {
        const pid = compiled.placeId(spec.place);
        mask[pid >>> WORD_SHIFT]! |= (1 << (pid & BIT_MASK));
      }
      inputPlaceMaskWords[tid] = mask;
    }
    this.inputPlaceMaskWords = inputPlaceMaskWords;
  }

  // ==================== Factory Methods ====================

  static compile(net: PetriNet): PrecompiledNet {
    return new PrecompiledNet(CompiledNet.compile(net));
  }

  static compileFrom(compiled: CompiledNet): PrecompiledNet {
    return new PrecompiledNet(compiled);
  }

  // ==================== Sparse Enablement Check ====================

  /**
   * Three-case sparse enablement check:
   * 1. Empty mask (needsSingleWordIndex == -2): always passes
   * 2. Single-word mask (>=0): one comparison
   * 3. Multi-word mask (-1): iterate precomputed sparse indices
   */
  canEnableSparse(tid: number, snapshot: Uint32Array): boolean {
    // Check needs mask
    const needsIdx = this.needsSingleWordIndex[tid]!;
    if (needsIdx === SPARSE_EMPTY) {
      // No needs — passes
    } else if (needsIdx >= 0) {
      // Single word
      const m = this.needsSingleWordMask[tid]!;
      if ((snapshot[needsIdx]! & m) !== m) return false;
    } else {
      // Multi-word sparse
      const indices = this.needsSparseIndices[tid]!;
      const masks = this.needsSparseMasks[tid]!;
      for (let i = 0; i < indices.length; i++) {
        const m = masks[i]!;
        if ((snapshot[indices[i]!]! & m) !== m) return false;
      }
    }

    // Check inhibitor mask
    const inhIdx = this.inhibitorSingleWordIndex[tid]!;
    if (inhIdx === SPARSE_EMPTY) {
      return true; // No inhibitors
    } else if (inhIdx >= 0) {
      return (snapshot[inhIdx]! & this.inhibitorSingleWordMask[tid]!) === 0;
    } else {
      const indices = this.inhibitorSparseIndices[tid]!;
      const masks = this.inhibitorSparseMasks[tid]!;
      for (let i = 0; i < indices.length; i++) {
        if ((snapshot[indices[i]!]! & masks[i]!) !== 0) return false;
      }
      return true;
    }
  }
}

// ==================== Compilation Helpers ====================

function compileSparse(
  mask: Uint32Array,
  wordCount: number,
  singleWordIndex: Int8Array,
  singleWordMask: Uint32Array,
  sparseIndices: number[][],
  sparseMasks: number[][],
  tid: number,
): void {
  let nonZeroCount = 0;
  let lastNonZeroWord = -1;

  for (let w = 0; w < wordCount; w++) {
    if (mask[w] !== 0) {
      nonZeroCount++;
      lastNonZeroWord = w;
    }
  }

  if (nonZeroCount === 0) {
    singleWordIndex[tid] = SPARSE_EMPTY;
    singleWordMask[tid] = 0;
    sparseIndices[tid] = [];
    sparseMasks[tid] = [];
  } else if (nonZeroCount === 1) {
    singleWordIndex[tid] = lastNonZeroWord;
    singleWordMask[tid] = mask[lastNonZeroWord]!;
    sparseIndices[tid] = [];
    sparseMasks[tid] = [];
  } else {
    singleWordIndex[tid] = SPARSE_MULTI;
    singleWordMask[tid] = 0;
    const idx: number[] = [];
    const msk: number[] = [];
    for (let w = 0; w < wordCount; w++) {
      if (mask[w] !== 0) {
        idx.push(w);
        msk.push(mask[w]!);
      }
    }
    sparseIndices[tid] = idx;
    sparseMasks[tid] = msk;
  }
}

function compileConsumeProgram(t: Transition, compiled: CompiledNet): number[] {
  const ops: number[] = [];

  for (const spec of t.inputSpecs) {
    const pid = compiled.placeId(spec.place);
    switch (spec.type) {
      case 'one':
        ops.push(CONSUME_ONE, pid);
        break;
      case 'exactly':
        ops.push(CONSUME_N, pid, spec.count);
        break;
      case 'all':
        ops.push(CONSUME_ALL, pid);
        break;
      case 'at-least':
        ops.push(CONSUME_ATLEAST, pid, spec.minimum);
        break;
    }
  }

  for (const arc of t.resets) {
    const pid = compiled.placeId(arc.place);
    ops.push(RESET, pid);
  }

  return ops;
}

function compileReadProgram(t: Transition, compiled: CompiledNet): number[] {
  const pids: number[] = [];
  for (const arc of t.reads) {
    pids.push(compiled.placeId(arc.place));
  }
  return pids;
}

function simpleOutputPlace(spec: Out, compiled: CompiledNet): number {
  if (spec.type === 'place') {
    return compiled.placeId(spec.place);
  }
  return -1; // complex
}
